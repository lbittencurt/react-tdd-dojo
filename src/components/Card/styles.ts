import styled from 'styled-components'

export const Card = styled.article`
  width: 30rem;
  height: 18rem;
  background: #ffffff;
  box-shadow: 2px 2px 4px rgba(0, 0, 0, 0.07);
  border-radius: 4px;
  padding: 2.6rem;
  margin: 4rem;
`

export const Title = styled.h1`
  font-weight: bold;
  font-size: 24px;
  color: #4a4a4a;
`

export const Subtitle = styled.h2`
  font-weight: normal;
  font-size: 16px;
  color: #797979;
  margin-bottom: 3rem;
`

export const Button = styled.button`
  background: #456fe8;
  border-radius: 5rem;
  padding: 0.8rem 2.2rem;
  color: #ffffff;
  font-weight: bold;
  font-size: 1.8rem;
  border: 0;
  cursor: pointer;

  :hover {
    background: #3652a4;
    transition: background 0.4s;
  }
`
