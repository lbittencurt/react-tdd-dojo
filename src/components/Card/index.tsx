import * as S from './styles'
import { useState } from 'react'

const Card = () => {
  const [buttonLabel, setButtonLabel] = useState('Create')

  const submit = () => {
    setButtonLabel('Creating...')
  }

  return (
    <S.Card role="card">
      <S.Title>React component</S.Title>
      <S.Subtitle>How to create a component</S.Subtitle>
      <S.Button type="submit" onClick={submit}>
        {buttonLabel}
      </S.Button>
    </S.Card>
  )
}

export default Card
