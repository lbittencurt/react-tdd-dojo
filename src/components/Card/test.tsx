import { render, screen, fireEvent } from '@testing-library/react'

import Card from '.'

describe('Card', () => {
  //title [OK]
  it('Should card render heading title', () => {
    render(<Card />)
    expect(
      screen.getByRole('heading', { name: /react component/i })
    ).toBeInTheDocument()
  })

  //subtitle [OK]
  it('Should card render heading subtitle', () => {
    render(<Card />)
    expect(
      screen.getByRole('heading', { name: /how to create a component/i })
    ).toBeInTheDocument()
  })

  //button [OK]
  it('Should card render an submit button', () => {
    render(<Card />)
    expect(screen.getByRole('button', { name: /create/i })).toBeInTheDocument()
  })

  //card
  it('Should card render article with card role', () => {
    render(<Card />)
    expect(screen.getByRole('card')).toBeInTheDocument()
  })

  //style [OK]
  it('Should card render and take snapshot', () => {
    const { container } = render(<Card />)
    expect(container.firstChild).toMatchSnapshot()
  })
  it('Should card button has primary color', () => {
    const SUBMIT_BUTTON_COLOR = '#456FE8'

    render(<Card />)
    const buttonContainer = screen.getByRole('button', { name: /create/i })
    expect(buttonContainer).toHaveStyle({
      'background-color': SUBMIT_BUTTON_COLOR
    })
  })

  //logic [OK]
  it('Should button change text to "creating..." when clicked', () => {
    render(<Card />)
    const buttonContainer = screen.getByRole('button', { name: /create/i })

    fireEvent.click(buttonContainer)

    expect(buttonContainer).toHaveTextContent('Creating')
  })
})
