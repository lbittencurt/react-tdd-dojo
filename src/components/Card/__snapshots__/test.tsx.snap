// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Should card render and take snapshot 1`] = `
.c0 {
  width: 30rem;
  height: 18rem;
  background: #ffffff;
  box-shadow: 2px 2px 4px rgba(0,0,0,0.07);
  border-radius: 4px;
  padding: 2.6rem;
  margin: 4rem;
}

.c1 {
  font-weight: bold;
  font-size: 24px;
  color: #4a4a4a;
}

.c2 {
  font-weight: normal;
  font-size: 16px;
  color: #797979;
  margin-bottom: 3rem;
}

.c3 {
  background: #456fe8;
  border-radius: 5rem;
  padding: 0.8rem 2.2rem;
  color: #ffffff;
  font-weight: bold;
  font-size: 1.8rem;
  border: 0;
  cursor: pointer;
}

.c3:hover {
  background: #3652a4;
  -webkit-transition: background 0.4s;
  transition: background 0.4s;
}

<article
  class="c0"
  role="card"
>
  <h1
    class="c1"
  >
    React component
  </h1>
  <h2
    class="c2"
  >
    How to create a component
  </h2>
  <button
    class="c3"
    type="submit"
  >
    Create
  </button>
</article>
`;
